### PR TITLE
model: Minicpmo

### PR DESCRIPTION
**Pull request by mickqian**
## Motivation

Support openbmb/MiniCPM-o-2_6

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html).

### CODE REVIEW

**Comment by mickqian** [comment]
The length input_ids from a prompt including video inputs(e.g., 30k+ tokens) would be split into several forwards, thus making the vision embedding (which relies on detecting the start and end of image_token_id) fail.

File: test/srt/test_vision_openai_server.py [hunk 1]
```diff
@@ -463,6 +480,91 @@ def setUpClass(cls):
...
+            other_args=[
+                "--trust-remote-code",
+                "--chat-template",
+                "minicpmo",
+                "--disable-cuda-graph",
+            ],
+        )
+        cls.base_url += "/v1"
+
+    def _test_audio_chat_completion(self):
```
Comment (mickqian) [inline test/srt/test_vision_openai_server.py]: will need a standard audio file to test the audio-understanding feature

File: python/sglang/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -251,6 +251,10 @@ def init_torch_distributed(self):
             dist_init_method = f"tcp://127.0.0.1:{self.dist_port}"
         set_custom_all_reduce(not self.server_args.disable_custom_all_reduce)
 
+        # Add cleanup of existing distributed environment
```
Comment (yizhang2077) [inline python/sglang/srt/model_executor/model_runner.py]: what would happen if we don't add this line here?

File: python/sglang/srt/models/minicpmv.py [hunk 3]
```diff
@@ -664,7 +667,7 @@ def __init__(
         super().__init__()
         # All MiniCPM-V models disable `tie_word_embeddings` but
         # `PretrainedConfig.tie_word_embeddings` defaults to True; we cannot
-        # check `tie_word_embeddings` until SGLang integrate MiniCPM-V model
+        # check `tie_word_embeddings` until vLLM integrate MiniCPM-V model
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmv.py]: why do we need revert this line to vLLM

File: python/sglang/srt/models/minicpmv.py [hunk 4]
```diff
@@ -964,10 +934,51 @@ def get_vision_embedding(
         raise NotImplementedError
 
     def get_vision_hidden_states(self, data: MiniCPMVImageInputs) -> torch.Tensor:
-        raise NotImplementedError
+        print("vision embedding")
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmv.py]: remove print here

File: python/sglang/srt/models/minicpmo.py [hunk 5]
```diff
@@ -0,0 +1,2110 @@
+# Adapted from
+# https://github.com/huggingface/transformers/blob/v4.28.0/src/transformers/models/llama/modeling_llama.py
+# Copyright 2023 The vLLM team.
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmo.py]: It seems we only refer https://huggingface.co/openbmb/MiniCPM-o-2_6/blob/main/modeling_minicpmo.py? I think we need change copyright and  reference here

**Comment by yizhang2077** [comment]
I think we should add test `.mp3` files to other places instead of adding to project directly.

**Comment by zhaochenyang20** [comment]
@mickqian Could you solve the conflicts? Thanks!

**Comment by zhaochenyang20** [comment]
@mickqian rebase with the main

**Comment by mickqian** [comment]
> I think we should add test `.mp3` files to other places instead of adding to project directly.我认为我们应该将测试 .mp3 文件添加到其他地方，而不是直接添加到项目中。

Any suggestions?

**Comment by zhaochenyang20** [comment]
> > I think we should add test `.mp3` files to other places instead of adding to project directly.我认为我们应该将测试 .mp3 文件添加到其他地方，而不是直接添加到项目中。
> 
> Any suggestions?

For examples, you can create a repo like `mick-sglang-material`, and download it from your repo.

**Comment by yizhang2077** [comment]
format it

**Comment by WangVertex** [comment]
Does it support voice output?

**Comment by zhaochenyang20** [comment]
@mickqian yeah. fix the conflicts and we can merge it after yi's approval @yizhang2077

File: test/srt/test_vision_llm.py [hunk 6]
```diff
@@ -199,12 +235,102 @@ async def test_encode_output(self):
...
+            and obj != VisionLLMLogitsBase
+        ):
+            test_classes.append(obj)
+    return test_classes
+
+
 if __name__ == "__main__":
-    unittest.main()
+
+    # TODO: Use standard unittest
```
Comment (yizhang2077) [inline test/srt/test_vision_llm.py]: Why do we need change unittest here?

File: python/sglang/srt/managers/image_processor.py [hunk 7]
```diff
@@ -1,55 +0,0 @@
...
-    for model_cls, processor_cls in IMAGE_PROCESSOR_MAPPING.items():
-        if model_cls.__name__ in hf_config.architectures:
-            return processor_cls(hf_config, server_args, processor)
-    raise ValueError(
-        f"No image processor found for architecture: {hf_config.architectures}"
-    )
-
-
-def get_dummy_image_processor():
-    return DummyImageProcessor()
```
Comment (yizhang2077) [inline python/sglang/srt/managers/image_processor.py:34]: why remove these files?

File: test/srt/test_vision_llm.py (hunk 6 above)
Comment (mickqian) [reply test/srt/test_vision_llm.py]: if not changed, a error about distributed  env would be raised. Unlike tests with openai servers, model_runner can't easily cleaned and reinitialized in a single thread. I'm still trying to figure out a better way to deal with logits test.

**Comment by zhaochenyang20** [comment]
thanks. I will merge transformers 4.48.3 today.

**Comment by zhaochenyang20** [comment]
@mickqian since the refactor is done, could we detach this PR of minicpm alone and only merge minicpm in this PR?

**Comment by zhaochenyang20** [comment]
@mickqian I've asked OpenBMB for help. They will help to make a new PR on minicpm-o vision and tts.

```bash
LLM prefill: **[fast] + decode [fast]**

MLLM: **{modal process [depends on your choice] encoder = HF} + {prefill [fast] + decode [fast]}**

TTS: decoder {sequence -> prefill -> decode = token ids} => decoder layer
```

**Comment by mickqian** [comment]
Depends on #4661

**Comment by zhaochenyang20** [comment]
great work mick!

File: test/srt/test_vision_openai_server.py [hunk 8]
```diff
@@ -272,21 +275,18 @@ def test_video_chat_completion(self):
...
         messages = self.prepare_video_messages(file_path)
 
-        video_request = client.chat.completions.create(
+        response = client.chat.completions.create(
             model="default",
             messages=messages,
             temperature=0,
             max_tokens=1024,
-            stream=True,
+            stream=False,
```
Comment (yizhang2077) [inline test/srt/test_vision_openai_server.py:283]: why change stream False here?

File: python/sglang/srt/utils.py [hunk 9]
```diff
@@ -507,9 +506,67 @@ def decode_video_base64(video_base64):
...
+        audio = resample(audio, num_samples)
+
+    # Convert to mono if requested and audio is stereo
+    if mono and len(audio.shape) > 1:
+        audio = np.mean(audio, axis=1)
+
+    return audio
+
+
+def load_audio(audio_file: str, sr: int = 16000, mono: bool = True) -> np.ndarray:
```
Comment (yizhang2077) [inline python/sglang/srt/utils.py]: why there are two `load_audio` funcs here?

File: python/sglang/srt/managers/processors/base_processor.py [hunk 10]
```diff
@@ -4,16 +4,16 @@
 import multiprocessing as mp
 import os
```
Comment (yizhang2077) [inline python/sglang/srt/managers/processors/base_processor.py]: Can we rename the folder `processors` as `multimodal_prcoessors` ?

File: python/sglang/srt/managers/schedule_batch.py [hunk 11]
```diff
@@ -144,11 +144,11 @@ def to_json(self):
 
 
 @dataclasses.dataclass
-class ImageInputs:
+class MultimodalInputs:
     """The image related inputs."""
 
     pixel_values: Union[torch.Tensor, np.array]
-    image_hashes: Optional[list] = None
+    data_hashes: Optional[list] = None
```
Comment (yizhang2077) [inline python/sglang/srt/managers/schedule_batch.py:151]: can we rename it as `mm_hashes` or `multimodal_hashes`?

**Comment by zhaochenyang20** [comment]
@mickqian is this ready?

File: python/sglang/srt/managers/processors/minicpm.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                audios,
+            )
+        else:
+            multimodal_data_inputs = self._processor(
+                images=images, text=input_text, audios=audios, return_tensors="pt"
+            )
+
+        return multimodal_data_inputs
+
+    async def process_data_async(
```
Comment (yizhang2077) [inline python/sglang/srt/managers/processors/minicpm.py]: can we rename it as `process_mm_data_async`?

File: python/sglang/srt/managers/processors/llava.py [hunk 13]
```diff
@@ -14,7 +14,7 @@
 from sglang.utils import get_exception_traceback
 
 
-class LlavaImageProcessor(BaseImageProcessor):
+class LlavaProcessor(BaseProcessor):
```
Comment (yizhang2077) [inline python/sglang/srt/managers/processors/llava.py]: Keep `LlavaImageProcessor` here

**Comment by zhaochenyang20** [comment]
I go over the codes change, it seems reasonalbe. but I don't get why there could be 3000+ lines of codes changed. is it expected?

File: python/sglang/srt/managers/processors/base_processor.py [hunk 14]
```diff
@@ -24,21 +24,41 @@ def get_global_processor():
...
+
+    def collect(self) -> list[str]:
+        return [
+            token
+            for token in [self.image_token, self.video_token, self.audio_token]
+            if token
+        ]
+
+
+class BaseProcessor(ABC):
```
Comment (yizhang2077) [inline python/sglang/srt/managers/processors/base_processor.py]: can we rename it as `BaseMultiModalProcessor` (vllm uses this)

**Review by yizhang2077 — COMMENTED** [review]
continue reviewing...

**Comment by mickqian** [comment]
> 3000

some rename ( image -> multi modal data) is involved. Some of them can be done in another pr

**Comment by zhaochenyang20** [comment]
> > 3000
> 
> some rename ( image -> multi modal data) is involved. Some of them can be done in another pr

no worry. we can do it here

File: python/sglang/srt/models/minicpmo.py [hunk 15]
```diff
@@ -0,0 +1,1993 @@
+# Copyright 2023-2024 SGLang Team
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmo.py:3]: Add reference here:https://huggingface.co/openbmb/MiniCPM-o-2_6/blob/main/modeling_minicpmo.py

File: python/sglang/srt/models/minicpmo.py [hunk 16]
```diff
@@ -0,0 +1,1993 @@
...
+            )
+            if audio_bounds.numel() == 0:
+                input_embeds = input_embeds.squeeze(0)
+                # TODO
+                logger.warn("Unimplemented logic. Please try disabling chunked prefill")
+                return input_embeds
+            audio_bounds = audio_bounds.unsqueeze(0)
+            bs = len(input_embeds)
+
+            if stream_input:
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmo.py:1795]: It seems we don't support stream_input here, can we remove it?

File: python/sglang/srt/models/minicpmo.py [hunk 17]
```diff
@@ -0,0 +1,1993 @@
...
+        """
+        input_lengths_after_cnn = (input_lengths - 1) // 2 + 1
+        input_lengths_after_pooling = (
+            input_lengths_after_cnn - self.config.audio_pool_step
+        ) // self.config.audio_pool_step + 1
+        input_lengths_after_pooling = input_lengths_after_pooling.to(dtype=torch.int32)
+
+        return input_lengths_after_cnn, input_lengths_after_pooling
+
+    def get_audio_embedding_streaming(self, multimodal_input: MultimodalInputs):
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmo.py:1540]: It seems we do not support streaming, can we remove it?

File: python/sglang/srt/models/minicpmo.py [hunk 18]
```diff
@@ -0,0 +1,1995 @@
...
+
+        # init tts module
+        self.config.init_tts = False
+        logger.info("TTS is disabled for now")
+        if self.config.init_tts:
+            # print("tts enabled")
+            assert (
+                _tts_deps
+            ), "please make sure vector_quantize_pytorch and vocos are installed."
+            self.tts = self.init_tts_module()
```
Comment (yizhang2077) [inline python/sglang/srt/models/minicpmo.py:1453]: `self.tts` seems used in forward, which means a lot of code can be removed?

File: test/srt/test_vlm_accuracy.py [hunk 19]
```diff
@@ -208,5 +246,74 @@ async def test_vlm_embedding_output(self):
...
+            and issubclass(obj, VisionLLMLogitsBase)
+            and obj != VisionLLMLogitsBase
+        ):
+            test_classes.append(obj)
+    return test_classes
+
+
 if __name__ == "__main__":
-    unittest.main()
+
```
Comment (yizhang2077) [inline test/srt/test_vlm_accuracy.py]: why we change unnitest.main here?

File: test/srt/test_vlm_accuracy.py [hunk 20]
```diff
@@ -208,5 +246,74 @@ async def test_vlm_embedding_output(self):
...
+try:
+    import librosa
+
+    librosa_installed = True
+except:
+    librosa_installed = False
+
+if librosa_installed:
+
+    class TestMiniCPMOLogits(TestMiniCPMVLogits):
```
Comment (yizhang2077) [inline test/srt/test_vlm_accuracy.py]: can we install librosa in `ci_install_dependency.sh` to open this test? @zhyncs

File: test/srt/test_vision_openai_server.py [hunk 21]
```diff
@@ -392,6 +393,77 @@ def test_mixed_batch(self):
...
+        print("-" * 30)
+
+        audio_response = audio_response.lower()
+
+        self.assertIsNotNone(audio_response)
+        self.assertGreater(len(audio_response), 0)
+
+        return audio_response
+
+    def _test_audio_speech_completion(self):
```
Comment (yizhang2077) [inline test/srt/test_vision_openai_server.py:442]: Can we directly move audio related test into `TestMinicpmoServer`?

**Comment by yizhang2077** [comment]
LGTM，after ci passes and unit test related issues have been solved, this PR can be merged. cc @zhaochenyang20 and plz paste MMMU benchmark result here~ @mickqian

**Comment by yizhang2077** [comment]
FYI, tts model in current Minicpmo is not used, as well as streaming audio embedding

**Comment by mickqian** [comment]
> LGTM，after ci passes and unit test related issues have been solve, this PR can be merged. cc @zhaochenyang20 and plz paste MMMU benchmark result here~   @mickqian


|     |  accuracy  |   
| --- | --- | 
|  sglang  |  0.481     |
|   hf |  0.49    |

**Comment by zhaochenyang20** [comment]
@yizhang2077 wait for yi's approval!

**Review by yizhang2077 — APPROVED** [review]

**Comment by zhaochenyang20** [comment]
great！let me rerun the CI and merge it asap

**Comment by yizhang2077** [comment]
@zhaochenyang20 we may need contributor to support audio benchmark

**Comment by zhaochenyang20** [comment]
great work1!

**Review by merrymercy — COMMENTED** [review]
Please move all model-specific files under `python/sglang/srt/managers/multimodal_processors` to `python/sglang/srt/models/`
